group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetAndMetadata.java

```diff
@@ -114,12 +116,29 @@ public static OffsetAndMetadata fromRecord(
     ) {
         return new OffsetAndMetadata(
             record.offset(),
-            record.leaderEpoch() == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
-                OptionalInt.empty() : OptionalInt.of(record.leaderEpoch()),
+            ofSentinel(record.leaderEpoch()),
             record.metadata(),
             record.commitTimestamp(),
-            record.expireTimestamp() == OffsetCommitRequest.DEFAULT_TIMESTAMP ?
-                OptionalLong.empty() : OptionalLong.of(record.expireTimestamp())
+            ofSentinel(record.expireTimestamp())
+        );
+    }
+
+    /**
+     * @return An OffsetAndMetadata created from an OffsetCommitRequestPartition request.
+     */
+    public static OffsetAndMetadata fromRequest(
+        OffsetCommitRequestData.OffsetCommitRequestPartition partition,
+        long currentTimeMs,
+        OptionalLong expireTimestampMs
+    ) {
+        return new OffsetAndMetadata(
+            partition.committedOffset(),
+            ofSentinel(partition.committedLeaderEpoch()),
+            partition.committedMetadata() == null ?
+                OffsetAndMetadata.NO_METADATA : partition.committedMetadata(),
+            partition.commitTimestamp() == OffsetCommitRequest.DEFAULT_TIMESTAMP ?
```

Comment:
Just for my understanding, if we have no timestamp/default in the request we use the current time, but in the record we just set the sentinel/empty?

Reply:
In the record, we store the decided time (either current time or the provided one). We store a sentinel (-1) for the expire timestamp if not set though.

Reply:
I think I was reading the diff wrong again -- I see it correctly now.